extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/HttpUpgradePolicy.java

```diff
@@ -0,0 +1,57 @@
+package io.quarkus.websockets.next;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * A policy that controls which requests are allowed to upgrade the HTTP connection to a WebSocket connection.
+ * CDI beans implementing this interface are invoked on every request.
+ */
+public interface HttpUpgradePolicy {
+
+    /**
+     * This method inspects HTTP Upgrade context and either allows or denies upgrade to a WebSocket connection;
+     * CDI beans implementing this method must never use {@link RoutingContext} to send response directly.
+     *
+     * @param context {@link HttpUpgradeContext}
+     * @return check result; must never be null
+     */
+    Uni<CheckResult> checkUpgrade(HttpUpgradeContext context);
+
+    /**
+     * @param routingContext {@link RoutingContext}
+     * @param securityIdentity {@link SecurityIdentity}; the identity is null if the Quarkus Security extension is absent
+     */
+    record HttpUpgradeContext(RoutingContext routingContext, SecurityIdentity securityIdentity) {
```

Comment:
I wonder if we should start with `io.vertx.core.http.HttpServerRequest` instead of `RoutingContext` and then add more "possibilities" if needed...

Reply:
I actually let there `RoutingContext` as I thought users may want to set response headers, but you propose an option to add response headers to rejected result, which I like better.

I agree it should be `io.vertx.core.http.HttpServerRequest` because I don't want users to manipulate with resonse like call `ctx.responde.end()` etc. I'll replace it.

Reply:
there is more than one `HttpServerRequest` wrapper, I am not able to determine if it can happen that one of the wrapper won't come in. Let's work against the `HttpServerRequest` interface and document that it is HTTP 1.X.

Reply:
You should work with `io.vertx.core.http.HttpServerRequest` here. HTTP 1.x is implicit because AFAIK it's not possible to upgrade an HTTP2 request...